_posts/2023/2023-02-27-consider-find_each-for-active-record-assocation-loops.md

```diff
@@ -0,0 +1,62 @@
+---
+title: "Consider find_each for looping through Active Record associations"
+description: "Rarely used, but protects you against foot shooting"
+layout: article
+category: ruby
+image:
+  base: '2023/consider-find_each-for-looping-through-active-record-associations'
+  alt: "Light loops"
+  credit: "Nareeta Martin"
+  source: "https://unsplash.com/photos/za2gMSfLmzU"
+
+---
+
+The standard method for enumerating through groups of objects is `each`, both through arrays in Ruby and through Active Record models in Rails.
+
+However, if you are looping over all the records for a model (perhaps to backfill data) or even over a large association you can encounter severe memory and speed issues both loading and processing enormous amounts of records as you loop.
+
+You should consider the functionality provided by [Active Record’s Batches](https://api.rubyonrails.org/classes/ActiveRecord/Batches.html), we’ll demonstrate using `find_each`.
+
+
+## Instead of…
+
+…looping through lots of Active Record objects using `.each`:
+
+```ruby
+post.comments.each do |comment|
+  # Do stuff with each comment: enqueue a job
+end
+```
+
+
+## Use…
+
+…use `.find_each` to more efficiently load records from the database:
+
+```ruby
+post.comments.find_each do |comment|
+  # Do stuff with each comment: enqueue a job?
+end
+```
+
+
+## Why?
+
+Using `each` makes _one_ SQL call to the database and tries to load the entire set of objects into memory and then loop over them. It’s the same as calling `post.comments.all.each` in the first example.
+
+This is problem in two dimensions. Firstly the database query may take a long time to execute or may timeout. Secondly, when (and if) it does, there’s likely to be significant memory usage as it loads all the records into memory in order to loop over them. 
+
+Using `.find_each` makes a series of more efficient SQL queries (with a bunch of sensible defaults) and uses a cursor-based approach to retrieve records from the database, which can be more efficient than loading all records into memory at once.
```

Comment:
Don't know what 'cursor-based approach' means. Is it commonly-understood?

*which is _often_ (?) much/a lot more efficient than...

Reply:
```suggestion
Using `.find_each` makes a series of more efficient SQL queries (with a bunch of sensible defaults) to retrieve records from the database, which is often a lot more efficient than loading all records into memory at once.
```